lightning/src/ln/channelmanager.rs

```diff
@@ -1256,11 +1261,23 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 			let channel_state = &mut *channel_state_lock;
 			match channel_state.by_id.entry(channel_id.clone()) {
 				hash_map::Entry::Occupied(mut chan_entry) => {
-					let (shutdown_msg, failed_htlcs) = chan_entry.get_mut().get_shutdown()?;
+					let counterparty_node_id = chan_entry.get().get_counterparty_node_id();
+					let their_features = {
+						let per_peer_state = self.per_peer_state.read().unwrap();
+						let peer_state = per_peer_state.get(&counterparty_node_id).unwrap().lock().unwrap();
+						peer_state.latest_features.clone()
+					};
+					let (shutdown_msg, monitor_update, failed_htlcs) = chan_entry.get_mut().get_shutdown(&self.keys_manager, &their_features)?;
 					channel_state.pending_msg_events.push(events::MessageSendEvent::SendShutdown {
-						node_id: chan_entry.get().get_counterparty_node_id(),
+						node_id: counterparty_node_id,
 						msg: shutdown_msg
 					});
+					if let Some(monitor_update) = monitor_update {
+						if let Err(_) = self.chain_monitor.update_channel(chan_entry.get().get_funding_txo().unwrap(), monitor_update) {
+							// TODO: How should this be handled?
```

Comment:
Should be handled like any other error - call `handle_monitor_err` to convert the error into a `MsgHandleErrInternal`, then store it in a variable outside the lock (like `failed_htlcs` and `chan_option`) and `handle_error` the error outside of the lock.

Reply:
Ok, this turned out to be a bit more tricky than I imagined. Confirming my understanding below.

Change is in 49614d8c5490dcf8f0e5695efa18e3d47ec55525. Let me know if that is what you had in mind and if there's a simpler way of going about this. Need to do the same for `internal_shutdown`.

> There are two errors - permanent ones and temporary ones. In the case of permenent, yes, we should force-close instead and send an `error` message,

These are handled by `handle_monitor_err` and `handle_error`, respectively, IIUC.

> for temporary, we should still send the shutdown.

What I don't quite understand is why we would send `shutdown` if we aren't certain that our `ChannelMonitor` has been updated with the shutdown script yet. Is it because if `TemporaryError` ever becomes a `PermanentError`, we wouldn't care about the shutdown script since we are force closing in that case?

Reply:
Rebased and added for `internal_shutdown`, too: https://github.com/rust-bitcoin/rust-lightning/pull/1019/commits/a02493003b5ef21969686f9186f4ce22f395cc46

Reply:
> Change is in 49614d8. Let me know if that is what you had in mind and if there's a simpler way of going about this. Need to do the same for internal_shutdown.

Nope, that's about right. These things are tricky to handle. I left a few specific comments on that commit.

> These are handled by handle_monitor_err and handle_error, respectively, IIUC.

That's the `MsgHandleErrInternal` vs monitor-error distinction, not the permanent-vs-temporary distinction, both of which are handled by `handle_monitor_err` (and friends)

> What I don't quite understand is why we would send shutdown if we aren't certain that our ChannelMonitor has been updated with the shutdown script yet. Is it because if TemporaryError ever becomes a PermanentError, we wouldn't care about the shutdown script since we are force closing in that case?

The Temporary->Permanent thing isn't a big deal, as you note. The real question is what happens if someone gives a Temporary update, and then takes longer to update their `ChannelMonitor` than we do doing the full `closing_signed` negotiation (because only *after* `closing_signed` negotiation could a transaction ever appear on chain with the script contained in the monitor update). Unlike most other monitor updates (at least `RevokeAndACK`), sending a `Shutdown` message isn't an immediate lock-in to a new states and the monitor doesn't need to know about it immediately, only before the closing transaction appears on chain.

If we wait for the update to complete here, then we have to add `Shutdown` resending, which I really don't want to do. An alternative, which is much easier, is to just wait for the monitor update to complete before ever sending a `closing_signed`, something I can shove in #985